src/pure/Queue.mo

```diff
@@ -6,51 +6,64 @@
 /// The input queue is left unchanged.
 ///
 /// Examples of use-cases:
-/// Queue (FIFO) by using `pushBack()` and `popFront()`.
-/// Stack (LIFO) by using `pushFront()` and `popFront()`.
+/// - Queue (FIFO) by using `pushBack()` and `popFront()`.
+/// - Stack (LIFO) by using `pushFront()` and `popFront()`.
+/// - Deque (double-ended queue) by using any combination of push/pop operations on either end.
 ///
-/// A Queue is internally implemented as two lists, a head access list and a (reversed) tail access list,
-/// that are dynamically size-balanced by splitting.
+/// A Queue is internally implemented as a real-time double-ended queue based on the paper
+/// "Real-Time Double-Ended Queue Verified (Proof Pearl)". The implementation maintains
+/// worst-case constant time `O(1)` for push/pop operations through gradual rebalancing steps.
 ///
 /// Construction: Create a new queue with the `empty<T>()` function.
 ///
-/// Note on the costs of push and pop functions:
-/// * Runtime: `O(1)` amortized costs, `O(n)` worst case cost per single call.
-/// * Space: `O(1)` amortized costs, `O(n)` worst case cost per single call.
-///
-/// `n` denotes the number of elements stored in the queue.
+/// Note that some operations that traverse the elements of the queue (e.g. `forEach`, `values`) preserve the order of the elements,
+/// whereas others (e.g. `map`, `contains`, `any`) do NOT guarantee that the elements are visited in any order.
+/// The order is undefined to avoid allocations, making these operations more efficient.
```

Comment:
I wonder if that's the right tradeoff to make - users might unwittingly rely on ordering. Would it be hard to preserve the order?

Reply:
It's not hard to code, but inefficient. Traversing in order requires `O(n)` allocations (because about a half of elements are in a list in reversed order). I think it would also be confusing for users that their `contains` check eats so much memory 😄

Reply:
Since there are no breaking changes related to that I've opted for 'undefined' order for these pure-like operations.
So that we could always change that to 'front-to-back' order in future if needed.
Users can always use the `Iter` module directly to get the desired order.
But it would be nice to ask others in @dfinity/languages what they would expect 😄 
I'm not sure what is the right way to go. Maybe removing these functions is the way to go, no confusion, but no functionality, just use `Iter`